File: app/src/main/java/com/amplifyframework/datastore/sample/LocalPresenter.java
package com.amplifyframework.datastore.sample;

import android.util.Pair;

import com.amplifyframework.datastore.generated.model.Post;

import java.util.Locale;

import io.reactivex.Observable;
import io.reactivex.disposables.CompositeDisposable;

final class LocalPresenter implements LocalPresentation.Presenter {
    private final LocalPresentation.PostInteractor interactor;
    private final LocalPresentation.View view;
    private final CompositeDisposable ongoingOperations;

    LocalPresenter(LocalPresentation.PostInteractor interactor, LocalPresentation.View view) {
        this.interactor = interactor;
        this.view = view;
        this.ongoingOperations = new CompositeDisposable();
    }

    @Override
    public void createLocalItems() {
        ongoingOperations.clear();
        ongoingOperations.add(
            interactor.createRandom().subscribe(createdPost ->
                view.displayLocalLogLine(LogLine.create(title(createdPost), details(createdPost)))
            )
        );
    }

    @Override
    public void updateLocalItems() {
        ongoingOperations.clear();
        ongoingOperations.add(
            interactor.list()
                .flatMap(interactor::updateAll)
                .flatMapObservable(Observable::fromIterable)
                .subscribe(updatedPost ->
                    view.displayLocalLogLine(LogLine.update(title(updatedPost), details(updatedPost)))
                )
        );
    }

    @Override
    public void listLocalItems() {
        ongoingOperations.clear();
        ongoingOperations.add(
            interactor.list()
                .flatMapObservable(Observable::fromIterable)
                .subscribe(foundPost ->
                    view.displayLocalLogLine(LogLine.query(title(foundPost), details(foundPost)))
                )
        );
    }

    @Override
    public void deleteLocalItems() {
        ongoingOperations.clear();
        ongoingOperations.add(
            interactor.list()
                .flatMap(items -> interactor.deleteAll(items).toSingleDefault(items))
                .flatMapObservable(Observable::fromIterable)
                .subscribe(deletedPost ->
                    view.displayLocalLogLine(LogLine.delete(title(deletedPost), details(deletedPost)))
                )
        );
    }

    @Override
    public void startSubscription() {
        ongoingOperations.clear();
        ongoingOperations.add(
            interactor.subscribe()
                .doOnSubscribe(token -> view.displayLocalLogLine(LogLine.subscription("Started.", token.toString())))
                .doOnDispose(() -> view.displayLocalLogLine(LogLine.subscription("Disposed.", "")))
                .subscribe(
                    nextData -> view.displayLocalLogLine(LogLine.subscription(title(nextData), details(nextData))),
                    failure -> view.displayLocalLogLine(LogLine.subscription("Failed.", failure.getMessage())),
                    () -> view.displayLocalLogLine(LogLine.subscription("Ended gracefully.", ""))
                )
        );
    }

    @Override
    public void clearLocalLog() {
        view.clearLocalLineItems();
    }

    @Override
    public void stopAllLocalActivities() {
        ongoingOperations.clear();
    }

    private static String title(Post post) {
        return post.getTitle();
    }

    private static String details(Post post) {
        return post.getId().substring(0, 7);
    }

    private static String title(Pair<Post, Modification> pair) {
        return pair.first.getTitle();
    }

    private static String details(Pair<Post, Modification> pair) {
        return String.format(Locale.US, "%s, %s", pair.second.name(), pair.first.getId());
    }
}

File: app/src/main/java/com/amplifyframework/datastore/sample/LocalPresenter.kt
package com.amplifyframework.datastore.sample

import android.util.Pair

import com.amplifyframework.datastore.generated.model.Post
import com.amplifyframework.datastore.sample.LocalPresentation.PostInteractor
import com.amplifyframework.datastore.sample.LocalPresentation.Presenter
import com.amplifyframework.datastore.sample.LocalPresentation.View

import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.functions.Consumer

internal class LocalPresenter(private val interactor: PostInteractor, private val view: View) : Presenter {
    private val ongoingOperations: CompositeDisposable = CompositeDisposable()

    override fun createLocalItems() {
        ongoingOperations.clear()
        ongoingOperations.add(interactor.createRandom().subscribe(
            Consumer { view.displayLocalLogLine(LogLine.create(title(it), details(it))) }
        ))
    }

    override fun updateLocalItems() {
        ongoingOperations.clear()
        ongoingOperations.add(interactor.list()
            .flatMap { interactor.updateAll(it) }
            .flatMapObservable { Observable.fromIterable(it) }
            .subscribe { view.displayLocalLogLine(LogLine.update(title(it), details(it))) }
        )
    }

    override fun listLocalItems() {
        ongoingOperations.clear()
        ongoingOperations.add(interactor.list()
            .flatMapObservable { Observable.fromIterable(it) }
            .subscribe { view.displayLocalLogLine(LogLine.query(title(it), details(it))) }
        )
    }

    override fun deleteLocalItems() {
        ongoingOperations.clear()
        ongoingOperations.add(interactor.list()
            .flatMap { interactor.deleteAll(it).toSingleDefault(it) }
            .flatMapObservable { Observable.fromIterable(it) }
            .subscribe { view.displayLocalLogLine(LogLine.delete(title(it), details(it))) }
        )
    }

    override fun startSubscription() {
        ongoingOperations.clear()
        ongoingOperations.add(interactor.subscribe()
            .doOnSubscribe { view.displayLocalLogLine(LogLine.subscription("Started", it.toString())) }
            .doOnDispose { view.displayLocalLogLine(LogLine.subscription("Disposed.", "")) }
            .subscribe(
                { view.displayLocalLogLine(LogLine.subscription(title(it), details(it))) },
                { view.displayLocalLogLine(LogLine.subscription("Failed.", it.message!!)) },
                { view.displayLocalLogLine(LogLine.subscription("Ended gracefully.", "")) }
            ))
    }

    override fun clearLocalLog() = view.clearLocalLineItems()

    override fun stopAllLocalActivities() = ongoingOperations.clear()

    companion object {
        private fun title(post: Post): String = post.title

        private fun details(post: Post): String = post.id.substring(0, 7)

        private fun title(pair: Pair<Post, Modification>): String = pair.first!!.title

        private fun details(pair: Pair<Post, Modification>): String =
            "${pair.second!!.name}, ${pair.first!!.id}"
    }
}
